src/compiler/irgen.jl

```diff
@@ -298,34 +299,32 @@ function replace_throw!(mod::LLVM.Module)
             if m != nothing && return_type(ft) == LLVM.VoidType(ctx)
                 ex = m.captures[1]
 
-                # create a function that prints the exception name, and exits
-                fn′ = "ptx_throw_$ex"
-                f′ = if haskey(functions(mod), fn′)
-                    functions(mod)[fn′]
-                else
-                    ft′ = LLVM.FunctionType(LLVM.VoidType(ctx))
-                    f′ = LLVM.Function(mod, fn′, ft′)
-                    let builder = Builder(ctx)
-                        entry = BasicBlock(f′, "entry", ctx)
-                        position!(builder, entry)
-
-                        desc = replace(ex, '_'=>' ')
-                        cuprintf!(builder, "ERROR: a $ex exception occurred during kernel execution$cuprintf_endline")
-                        call!(builder, trap)
-                        ret!(builder)
-
-                        dispose(builder)
-                    end
-                    f′
-                end
-
-                # replace uses of the original function
+                # replace uses of the original function with a call to the run-time
                 for use in uses(f)
                     call = user(use)
                     @assert isa(call, LLVM.CallInst)
                     let builder = Builder(ctx)
                         position!(builder, call)
-                        call!(builder, f′)
+
+                        # report the exception
+                        name = globalstring_ptr!(builder, String(ex), "exception")
+                        name = ptrtoint!(builder, name, Runtime.report_exception.llvm_types[1])
+                        call!(builder, Runtime.report_exception, [name])
+
+                        # report each frame
+                        let rt = Runtime.report_exception_frame
+                            bt = backtrace(call)
+                            for (i,frame) in enumerate(bt)
+                                idx = ConstantInt(rt.llvm_types[1], i)
```

Comment:
These are the static backtraces you mentioned? At some point I was looking at `cuda-gdb` trying to figure out what it was doing, but I forgot about it and didn't keep notes :(
I vaguely remember: https://github.com/NVIDIA/cuda-gdb/tree/cuda-toolkit-7.5/libcudacore as dealing with core dumps from cuda and having functionality like reading the elf image and the memory of the device

Reply:
Oh that's interesting! Need to have a better look.

The current backtrace generation `backtrace(::LLVM.Instruction)` is at https://github.com/JuliaGPU/CUDAnative.jl/blob/4b384644a6172b344b59300ae5129118822e7e6c/src/compiler/validation.jl#L73-L117 -- works ok-ish since we recompile function in the call tree for every kernel so we typically have only one caller.